pkg/etcdcli/etcdcli_pool_test.go

```diff
@@ -0,0 +1,430 @@
+		poolRecorder.pool.Return(client)
+	}
+
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumCachedClients+1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 1, poolRecorder.numCloseCalls)
+}
+
+func TestNewClientWithTickets(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+
+	// this should block and return an error
+	client, err := poolRecorder.pool.Get()
+	assert.Nil(t, client)
+	assert.Errorf(t, err, "too many active cache clients, rejecting to create new one")
+	// returning one should unlock the get again
+	poolRecorder.pool.Return(clients[0])
+	client, err = poolRecorder.pool.Get()
+	assert.NotNil(t, client)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)        // no new call added
+	assert.Equal(t, maxNumClientTickets+2, poolRecorder.numEndpointCalls) // called Get twice additionally
+	assert.Equal(t, maxNumClientTickets+1, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+}
+
+func TestClosesReturnTickets(t *testing.T) {
+	integration.BeforeTestExternal(t)
+	testServer := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 3})
+	defer testServer.Terminate(t)
+
+	poolRecorder := newTestPool(testServer)
+	var clients []*clientv3.Client
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+		clients = append(clients, client)
+	}
+
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numNewCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numEndpointCalls)
+	assert.Equal(t, maxNumClientTickets, poolRecorder.numHealthCalls)
+	assert.Equal(t, 0, poolRecorder.numCloseCalls)
+
+	// return all clients to fill the internal cache and cause five to close
+	for i := 0; i < maxNumClientTickets; i++ {
+		poolRecorder.pool.Return(clients[i])
+	}
+	assert.Equal(t, maxNumCachedClients, poolRecorder.numCloseCalls)
+
+	// now we should be able to get the full amount of tickets again
+	for i := 0; i < maxNumClientTickets; i++ {
+		client, err := poolRecorder.pool.Get()
+		require.NoError(t, err)
+		assert.NotNil(t, client)
+	}
+
+	// replenish the maxNumCachedClients that were closed earlier
+	assert.Equal(t, maxNumClientTickets+maxNumCachedClients, poolRecorder.numNewCalls)
```

Comment:
Can we add a check for the number of available tickets too just to make it explicit that available tickets are the capacity to create more clients.
```suggestion
	assert.Equal(t, maxNumClientTickets+maxNumCachedClients, poolRecorder.numNewCalls)
	assert.Equal(t, maxNumClientTickets-maxNumCachedClients, len(poolRecorder.pool.availableTickets))
```

Reply:
yep, in this case the tickets are zero as all clients are handed out:
```
	// no tickets are available anymore, as we have handed out all clients
	assert.Equal(t, 0, len(poolRecorder.pool.availableTickets))
```